src/Redist/Microsoft.ML.DnnImageFeaturizer.ModelRedist/Microsoft.ML.DnnImageFeaturizer.ModelRedist.proj

```diff
@@ -0,0 +1,106 @@
+    Url="https://aka.ms/mlnet-resources/image/ResNet50Onnx/ResNet50.onnx"
+    DestinationFile="$(ModelPlacementDir)\ResNet50Onnx\ResNet50.onnx"
+    DestinationDir="$(ModelPlacementDir)\ResNet50Onnx" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ResNet101ModelFile Include="$(ModelPlacementDir)\ResNet101Onnx\ResNet101.onnx"
+    Url="https://aka.ms/mlnet-resources/image/ResNet101Onnx/ResNet101.onnx"
+    DestinationFile="$(ModelPlacementDir)\ResNet101Onnx\ResNet101.onnx"
+    DestinationDir="$(ModelPlacementDir)\ResNet101Onnx" />
+  </ItemGroup>
+
+  <Target Name="DownloadAlexNetPrepModels" Inputs="@(AlexNetPreprocessFile)" Outputs="%(AlexNetPreprocessFile.DestinationFile)">
+    <Message Importance="High" Text="Downloading external model files... %(AlexNetPreprocessFile.DestinationFile)" />
+    <MakeDir Directories="$(ModelPlacementDir)" />
+    <MakeDir Directories="%(AlexNetPreprocessFile.DestinationDir)" />
+    <DownloadFilesFromUrl Items="@(AlexNetPreprocessFile)"/>
+  </Target>
+
+  <Target Name="DownloadResNetPrepModels" 
+          DependsOnTargets="DownloadAlexNetPrepModels" Inputs="@(ResNetPreprocessFile)" Outputs="%(ResNetPreprocessFile.DestinationFile)">
+    <Message Importance="High" Text="Downloading external model files... %(ResNetPreprocessFile.DestinationFile)" />
+    <MakeDir Directories="$(ModelPlacementDir)" />
+    <MakeDir Directories="%(ResNetPreprocessFile.DestinationDir)" />
+    <DownloadFilesFromUrl Items="@(ResNetPreprocessFile)"/>
+  </Target>
+
+  <Target Name="DownloadAlexNetModels" 
+          DependsOnTargets="DownloadResNetPrepModels" Inputs="@(AlexNetModelFile)" Outputs="%(AlexNetModelFile.DestinationFile)">
+    <Message Importance="High" Text="Downloading external model files... %(AlexNetModelFile.DestinationFile)" />
+    <MakeDir Directories="$(ModelPlacementDir)" />
+    <MakeDir Directories="%(AlexNetModelFile.DestinationDir)" />
+    <DownloadFilesFromUrl Items="@(AlexNetModelFile)"/>
+  </Target>
+
+  <Target Name="DownloadResNet18Models" 
+          DependsOnTargets="DownloadAlexNetModels" Inputs="@(ResNet18ModelFile)" Outputs="%(ResNet18ModelFile.DestinationFile)">
+    <Message Importance="High" Text="Downloading external model files... %(ResNet18ModelFile.DestinationFile)" />
+    <MakeDir Directories="$(ModelPlacementDir)" />
+    <MakeDir Directories="%(ResNet18ModelFile.DestinationDir)" />
+    <DownloadFilesFromUrl Items="@(ResNet18ModelFile)"/>
+  </Target>
+
+  <Target Name="DownloadResNet50Models" 
+          DependsOnTargets="DownloadResNet18Models" Inputs="@(ResNet50ModelFile)" Outputs="%(ResNet50ModelFile.DestinationFile)">
+    <Message Importance="High" Text="Downloading external model files... %(ResNet50ModelFile.DestinationFile)" />
+    <MakeDir Directories="$(ModelPlacementDir)" />
+    <MakeDir Directories="%(ResNet50ModelFile.DestinationDir)" />
+    <DownloadFilesFromUrl Items="@(ResNet50ModelFile)"/>
+  </Target>
+
+  <Target Name="DownloadResNet101Models" 
```

Comment:
You can clean up a lot of the duplication in this file by following the same pattern/design as [https://github.com/dotnet/machinelearning/blob/master/src/Redist/Microsoft.ML.TensorFlow.Redist/Microsoft.ML.TensorFlow.Redist.proj](https://github.com/dotnet/machinelearning/blob/master/src/Redist/Microsoft.ML.TensorFlow.Redist/Microsoft.ML.TensorFlow.Redist.proj).

Daisy-chaining these Targets doesn't really make sense because nothing about `DownloadResNet101Models` depends on `DownloadResNet50Models` from running.

So instead, I would do the following:

1. Define an Item (like `<DnnModelFile>`) that has properties for the URL and for where the file should be downloaded to.
2. You can have a single Target like `DownloadDnnModels`, which downloads all the items defined in `<DnnModelFile>` and places them to their output destination.

Reply:
The overarching issue I had when looking at this (which is also a concern for other bin\obj related comments you had): each of the models needs to be placed in a separate directory. The OnnxTransform has this limitation where each model file has to be in a dir all by itself for it to work. From my experiments copying files in batches, I could not get the folder structure to be preserved and the files would end up in the same directory. I might have been doing something wrong though, so please let me know if I can copy multiple of them while preserving inner folder structure (causing each one to be in its own dir).

---
In reply to: [234759995](https://github.com/dotnet/machinelearning/pull/1447#discussion_r234759995) [](ancestors = 234759995)

Reply:
You can still put them in separate directories, even if you clean this file up.

See [https://github.com/dotnet/machinelearning/blob/master/src/Redist/Microsoft.ML.TensorFlow.Redist/Microsoft.ML.TensorFlow.Redist.proj#L103-L113](https://github.com/dotnet/machinelearning/blob/master/src/Redist/Microsoft.ML.TensorFlow.Redist/Microsoft.ML.TensorFlow.Redist.proj#L103-L113) (where we specify the TargetPath) and [https://github.com/dotnet/machinelearning/blob/master/src/Redist/Microsoft.ML.TensorFlow.Redist/Microsoft.ML.TensorFlow.Redist.proj#L121](https://github.com/dotnet/machinelearning/blob/master/src/Redist/Microsoft.ML.TensorFlow.Redist/Microsoft.ML.TensorFlow.Redist.proj#L121) (where we use that TargetPath to copy the files to separate directories).

Take a good look at the TensorFlow.Redist project for how to structure this file. It makes this downloading code much cleaner.

---
In reply to: [235548279](https://github.com/dotnet/machinelearning/pull/1447#discussion_r235548279) [](ancestors = 235548279,234759995)

Reply:
Ok, fixed it.

---
In reply to: [236389503](https://github.com/dotnet/machinelearning/pull/1447#discussion_r236389503) [](ancestors = 236389503,235548279,234759995)